File: chatbot-saas-platform/backend/src/WorkflowService/Services/WorkflowService.cs
using Microsoft.EntityFrameworkCore;
using Shared.Application.Common.Interfaces;
using WorkflowService.Services;
using WorkflowService.Models;
using Shared.Domain.Entities;

namespace WorkflowService.Services;

public class WorkflowService : IWorkflowService
{
    private readonly IApplicationDbContext _context;
    private readonly IWorkflowDesignerService _designerService;
    private readonly ILogger<WorkflowService> _logger;

    public WorkflowService(
        IApplicationDbContext context,
        IWorkflowDesignerService designerService,
        ILogger<WorkflowService> logger)
    {
        _context = context;
        _designerService = designerService;
        _logger = logger;
    }

    public async Task<Workflow> CreateWorkflowAsync(CreateWorkflowRequest request, Guid tenantId, Guid userId)
    {
        try
        {
            var workflow = new Workflow
            {
                Id = Guid.NewGuid(),
                Name = request.Name,
                Description = request.Description,
                TenantId = tenantId,
                CreatedByUserId = userId,
                Status = WorkflowStatus.Draft,
                Definition = request.Definition ?? new WorkflowDefinition(),
                Variables = request.Variables,
                Tags = request.Tags,
                Trigger = request.Trigger ?? new WorkflowTrigger(),
                Settings = request.Settings ?? new WorkflowSettings(),
                TemplateId = request.TemplateId,
                CreatedAt = DateTime.UtcNow
            };

            if (request.Definition != null)
            {
                workflow.Definition = await _designerService.ValidateWorkflowDefinitionAsync(request.Definition);
            }

            _context.Set<Workflow>().Add(workflow);
            await _context.SaveChangesAsync();

            _logger.LogInformation("Workflow {WorkflowId} created for tenant {TenantId}", workflow.Id, tenantId);
            return workflow;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating workflow for tenant {TenantId}", tenantId);
            throw;
        }
    }

    public async Task<List<Workflow>> GetWorkflowsAsync(Guid tenantId, int page = 1, int pageSize = 20)
    {
        try
        {
            return await _context.Set<Workflow>()
                .Where(w => w.TenantId == tenantId)
                .OrderByDescending(w => w.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting workflows for tenant {TenantId}", tenantId);
            throw;
        }
    }

    public async Task<Workflow?> GetWorkflowAsync(Guid workflowId, Guid tenantId)
    {
        try
        {
            return await _context.Set<Workflow>()
                .FirstOrDefaultAsync(w => w.Id == workflowId && w.TenantId == tenantId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting workflow {WorkflowId} for tenant {TenantId}", workflowId, tenantId);
            throw;
        }
    }

    public async Task<bool> UpdateWorkflowAsync(Workflow workflow)
    {
        try
        {
            var existingWorkflow = await _context.Set<Workflow>()
                .FirstOrDefaultAsync(w => w.Id == workflow.Id);

            if (existingWorkflow == null)
                return false;

            workflow.Definition = await _designerService.ValidateWorkflowDefinitionAsync(workflow.Definition);

            existingWorkflow.Name = workflow.Name;
            existingWorkflow.Description = workflow.Description;
            existingWorkflow.Definition = workflow.Definition;
            existingWorkflow.Variables = workflow.Variables;
            existingWorkflow.Tags = workflow.Tags;
            existingWorkflow.Trigger = workflow.Trigger;
            existingWorkflow.Settings = workflow.Settings;
            existingWorkflow.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating workflow {WorkflowId}", workflow.Id);
            return false;
        }
    }

    public async Task<bool> DeleteWorkflowAsync(Guid workflowId, Guid tenantId)
    {
        try
        {
            var workflow = await GetWorkflowAsync(workflowId, tenantId);
            if (workflow == null)
                return false;

            _context.Set<Workflow>().Remove(workflow);
            await _context.SaveChangesAsync();
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting workflow {WorkflowId}", workflowId);
            return false;
        }
    }

    public async Task<bool> ActivateWorkflowAsync(Guid workflowId, Guid tenantId)
    {
        try
        {
            var workflow = await GetWorkflowAsync(workflowId, tenantId);
            if (workflow == null)
                return false;

            await _designerService.ValidateWorkflowDefinitionAsync(workflow.Definition);

            workflow.Status = WorkflowStatus.Active;
            workflow.IsActive = true;
            workflow.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error activating workflow {WorkflowId}", workflowId);
            return false;
        }
    }

    public async Task<bool> DeactivateWorkflowAsync(Guid workflowId, Guid tenantId)
    {
        try
        {
            var workflow = await GetWorkflowAsync(workflowId, tenantId);
            if (workflow == null)
                return false;

            workflow.Status = WorkflowStatus.Inactive;
            workflow.IsActive = false;
            workflow.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deactivating workflow {WorkflowId}", workflowId);
            return false;
        }
    }

    public async Task<Workflow> CloneWorkflowAsync(Guid workflowId, Guid tenantId, string newName)
    {
        try
        {
            var originalWorkflow = await GetWorkflowAsync(workflowId, tenantId);
            if (originalWorkflow == null)
                throw new ArgumentException("Workflow not found");

            var clonedWorkflow = new Workflow
            {
                Id = Guid.NewGuid(),
                Name = newName,
                Description = $"Clone of {originalWorkflow.Name}",
                TenantId = tenantId,
                CreatedByUserId = originalWorkflow.CreatedByUserId,
                Status = WorkflowStatus.Draft,
                Definition = originalWorkflow.Definition,
                Variables = new Dictionary<string, object>(originalWorkflow.Variables),
                Tags = new List<string>(originalWorkflow.Tags),
                Trigger = originalWorkflow.Trigger,
                Settings = originalWorkflow.Settings,
                CreatedAt = DateTime.UtcNow
            };

            _context.Set<Workflow>().Add(clonedWorkflow);
            await _context.SaveChangesAsync();

            return clonedWorkflow;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error cloning workflow {WorkflowId}", workflowId);
            throw;
        }
    }

    public async Task<List<Workflow>> GetWorkflowsByTagAsync(string tag, Guid tenantId)
    {
        try
        {
            return await _context.Set<Workflow>()
                .Where(w => w.TenantId == tenantId && w.Tags.Contains(tag))
                .OrderByDescending(w => w.CreatedAt)
                .ToListAsync();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting workflows by tag {Tag} for tenant {TenantId}", tag, tenantId);
            throw;
        }
    }

    public async Task<WorkflowStatistics> GetStatisticsAsync(Guid tenantId, DateTime? startDate = null, DateTime? endDate = null)
    {
        try
        {
            var start = startDate ?? DateTime.UtcNow.AddDays(-30);
            var end = endDate ?? DateTime.UtcNow;

            var workflows = await _context.Set<Workflow>()
                .Where(w => w.TenantId == tenantId)
                .ToListAsync();

            var executions = await _context.Set<WorkflowExecution>()
                .Where(e => e.TenantId == tenantId && e.StartedAt >= start && e.StartedAt <= end)
                .ToListAsync();

            var stepExecutions = await _context.Set<WorkflowStepExecution>()
                .Where(se => executions.Select(e => e.Id).Contains(se.WorkflowExecutionId))
                .ToListAsync();

            return new WorkflowStatistics
            {
                TotalWorkflows = workflows.Count,
                ActiveWorkflows = workflows.Count(w => w.IsActive),
                TotalExecutions = executions.Count,
                SuccessfulExecutions = executions.Count(e => e.Status == WorkflowExecutionStatus.Completed),
                FailedExecutions = executions.Count(e => e.Status == WorkflowExecutionStatus.Failed),
                SuccessRate = executions.Count > 0 ? (double)executions.Count(e => e.Status == WorkflowExecutionStatus.Completed) / executions.Count * 100 : 0,
                AverageExecutionTime = executions.Where(e => e.Duration.HasValue).Any() ? 
                    TimeSpan.FromMilliseconds(executions.Where(e => e.Duration.HasValue).Average(e => e.Duration!.Value.TotalMilliseconds)) : 
                    TimeSpan.Zero,
                StepTypeUsage = stepExecutions.GroupBy(se => se.StepType).ToDictionary(g => g.Key, g => g.Count()),
                ExecutionsByWorkflow = executions.GroupBy(e => e.WorkflowId).ToDictionary(g => g.Key.ToString(), g => g.Count()),
                PeriodStart = start,
                PeriodEnd = end
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting workflow statistics for tenant {TenantId}", tenantId);
            throw;
        }
    }
}

File: chatbot-saas-platform/frontend/tenant-dashboard/src/services/workflowService.ts
import { apiClient } from './api';
import type { ApiResponse } from './api';

export interface Workflow {
  id: string;
  name: string;
  description: string;
  status: 'Draft' | 'Active' | 'Inactive' | 'Archived' | 'Error';
  version: string;
  isActive: boolean;
  definition?: WorkflowDefinition;
  variables: Record<string, any>;
  tags: string[];
  trigger?: WorkflowTrigger;
  settings?: WorkflowSettings;
  executionCount: number;
  lastExecutedAt?: string;
  templateId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowDefinition {
  steps: WorkflowStep[];
  connections: WorkflowConnection[];
  layout: WorkflowLayout;
}

export interface WorkflowStep {
  id: string;
  name: string;
  type: WorkflowStepType;
  configuration: Record<string, any>;
  position: { x: number; y: number };
  inputPorts: string[];
  outputPorts: string[];
  isStartStep: boolean;
  isEndStep: boolean;
  condition?: WorkflowStepCondition;
}

export interface WorkflowConnection {
  id: string;
  sourceStepId: string;
  sourcePort: string;
  targetStepId: string;
  targetPort: string;
  condition?: WorkflowConnectionCondition;
}

export interface WorkflowLayout {
  width: number;
  height: number;
  zoom: number;
  viewportPosition: { x: number; y: number };
}

export interface WorkflowTrigger {
  type: WorkflowTriggerType;
  configuration: Record<string, any>;
  isEnabled: boolean;
}

export interface WorkflowSettings {
  maxRetries: number;
  timeout: number;
  enableLogging: boolean;
  enableNotifications: boolean;
  errorHandling: 'Stop' | 'Continue' | 'Retry' | 'Escalate';
  customSettings: Record<string, any>;
}

export interface WorkflowStepCondition {
  expression: string;
  variables: Record<string, any>;
}

export interface WorkflowConnectionCondition {
  expression: string;
  variables: Record<string, any>;
}

export type WorkflowStepType = 'Start' | 'End' | 'Action' | 'Condition' | 'Loop' | 'Parallel' | 'Wait' | 'HttpRequest' | 'EmailSend' | 'DatabaseQuery' | 'ScriptExecution' | 'UserTask' | 'ServiceTask' | 'Timer' | 'Gateway' | 'SubWorkflow';

export type WorkflowTriggerType = 'Manual' | 'Scheduled' | 'Event' | 'Webhook' | 'FileUpload' | 'DatabaseChange' | 'MessageReceived';

export interface WorkflowExecution {
  id: string;
  workflowId: string;
  status: 'Pending' | 'Running' | 'Completed' | 'Failed' | 'Cancelled' | 'Timeout';
  inputData: Record<string, any>;
  outputData: Record<string, any>;
  startedAt: string;
  completedAt?: string;
  duration?: number;
  errorMessage?: string;
  triggerSource?: string;
  stepExecutions: WorkflowStepExecution[];
}

export interface WorkflowStepExecution {
  id: string;
  stepId: string;
  stepName: string;
  stepType: WorkflowStepType;
  status: 'Pending' | 'Running' | 'Completed' | 'Failed' | 'Skipped' | 'Cancelled';
  inputData: Record<string, any>;
  outputData: Record<string, any>;
  startedAt: string;
  completedAt?: string;
  duration?: number;
  errorMessage?: string;
  retryCount: number;
}

export interface WorkflowTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  definition: WorkflowDefinition;
  defaultVariables: Record<string, any>;
  tags: string[];
  isPublic: boolean;
  tenantId?: string;
  usageCount: number;
  rating: number;
}

export interface WorkflowStatistics {
  totalWorkflows: number;
  activeWorkflows: number;
  totalExecutions: number;
  successfulExecutions: number;
  failedExecutions: number;
  successRate: number;
  averageExecutionTime: number;
  stepTypeUsage: Record<WorkflowStepType, number>;
  executionsByWorkflow: Record<string, number>;
  periodStart: string;
  periodEnd: string;
}

export interface CreateWorkflowData {
  name: string;
  description: string;
  definition?: WorkflowDefinition;
  variables?: Record<string, any>;
  tags?: string[];
  trigger?: WorkflowTrigger;
  settings?: WorkflowSettings;
  templateId?: string;
}

export interface UpdateWorkflowData extends Partial<CreateWorkflowData> {
  id: string;
}

export interface ExecuteWorkflowData {
  inputData?: Record<string, any>;
  triggerSource?: string;
  waitForCompletion?: boolean;
}

export class WorkflowService {
  static async getWorkflows(page = 1, pageSize = 20): Promise<ApiResponse<Workflow[]>> {
    return apiClient.get(`/workflows?page=${page}&pageSize=${pageSize}`);
  }

  static async getWorkflow(id: string): Promise<ApiResponse<Workflow>> {
    return apiClient.get(`/workflows/${id}`);
  }

  static async createWorkflow(data: CreateWorkflowData): Promise<ApiResponse<Workflow>> {
    return apiClient.post('/workflows', data);
  }

  static async updateWorkflow(data: UpdateWorkflowData): Promise<ApiResponse<void>> {
    return apiClient.put(`/workflows/${data.id}`, data);
  }

  static async deleteWorkflow(id: string): Promise<ApiResponse<void>> {
    return apiClient.delete(`/workflows/${id}`);
  }

  static async activateWorkflow(id: string): Promise<ApiResponse<void>> {
    return apiClient.post(`/workflows/${id}/activate`);
  }

  static async deactivateWorkflow(id: string): Promise<ApiResponse<void>> {
    return apiClient.post(`/workflows/${id}/deactivate`);
  }

  static async cloneWorkflow(id: string, newName: string): Promise<ApiResponse<Workflow>> {
    return apiClient.post(`/workflows/${id}/clone`, { newName });
  }

  static async executeWorkflow(id: string, data: ExecuteWorkflowData): Promise<ApiResponse<{ executionId: string }>> {
    return apiClient.post(`/workflows/${id}/execute`, data);
  }

  static async getWorkflowExecutions(workflowId: string, page = 1, pageSize = 20): Promise<ApiResponse<WorkflowExecution[]>> {
    return apiClient.get(`/workflows/${workflowId}/executions?page=${page}&pageSize=${pageSize}`);
  }

  static async getWorkflowExecution(executionId: string): Promise<ApiResponse<WorkflowExecution>> {
    return apiClient.get(`/workflows/executions/${executionId}`);
  }

  static async cancelExecution(executionId: string): Promise<ApiResponse<void>> {
    return apiClient.post(`/workflows/executions/${executionId}/cancel`);
  }

  static async retryExecution(executionId: string): Promise<ApiResponse<void>> {
    return apiClient.post(`/workflows/executions/${executionId}/retry`);
  }

  static async getStatistics(startDate?: string, endDate?: string): Promise<ApiResponse<WorkflowStatistics>> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    return apiClient.get(`/workflows/statistics?${params.toString()}`);
  }

  static async getAvailableStepTypes(): Promise<ApiResponse<WorkflowStepType[]>> {
    return apiClient.get('/workflows/designer/step-types');
  }

  static async validateWorkflowDefinition(definition: WorkflowDefinition): Promise<ApiResponse<{ isValid: boolean; definition: WorkflowDefinition; message: string }>> {
    return apiClient.post('/workflows/designer/validate', definition);
  }

  static async getTemplates(category?: string, publicOnly = true): Promise<ApiResponse<WorkflowTemplate[]>> {
    const params = new URLSearchParams();
    if (category) params.append('category', category);
    params.append('publicOnly', publicOnly.toString());
    return apiClient.get(`/workflows/templates?${params.toString()}`);
  }

  static async createWorkflowFromTemplate(templateId: string, workflowName: string): Promise<ApiResponse<Workflow>> {
    return apiClient.post(`/workflows/templates/${templateId}/create-workflow`, { workflowName });
  }
}
